newIDE/app/src/UI/EditorMosaic/index.js

```diff
@@ -197,36 +198,61 @@ export default class EditorMosaic extends React.Component<Props, State> {
         editorName => editors[editorName].type === 'secondary'
       );
       if (secondaryEditorName) {
-        this.setState({
-          mosaicNode: replaceNode(
-            this.state.mosaicNode,
-            secondaryEditorName,
-            editorName
-          ),
-        });
-
-        this._persistNodes();
+        this.setState(
+          {
+            mosaicNode: replaceNode(
+              this.state.mosaicNode,
+              secondaryEditorName,
+              editorName
+            ),
+          },
+          () => {
+            this._onOpenedEditorsChanged();
+          }
+        );
+
         return true;
       }
     }
 
-    // Open a new editor at the indicated position
-    this.setState({
-      mosaicNode: addNode(
-        this.state.mosaicNode,
-        editorName,
-        position,
-        splitPercentage,
-        direction
-      ),
-    });
+    // Open a new editor at the indicated position.
+    this.setState(
+      {
+        mosaicNode: addNode(
+          this.state.mosaicNode,
+          editorName,
+          position,
+          splitPercentage,
+          direction
+        ),
+      },
+      () => {
+        this._onOpenedEditorsChanged();
+      }
+    );
 
-    this._persistNodes();
     return true;
   };
 
+  getOpenedEditorNames = (): Array<string> => {
+    return getLeaves(this.state.mosaicNode);
+  };
+
   _onChange = (mosaicNode: EditorMosaicNode) => {
     this.setState({ mosaicNode });
+  };
+
+  _onChanged = (mosaicNode: EditorMosaicNode) => {
```

Comment:
Maybe you can re-use it in the functions above?

Reply:
(it's quite a confusing function though, it does the same thing as the onChange, but calls on more function. Should this replace the _onChange ?)

Reply:
Not really:

- onChange = update done 60 times a second when you drag a panel border for example. So it must call setState but NOT notify the parent of anything (to avoid a huge drop in performance because of 60 renders a second)
- onChanged = update done once when a change is finished (drag ended). So it must call setState and notify the parent that something happened (so we can update icons)

Reply:
(and we don't want to call setState in `_onOpenedEditorsChanged`, because it's called when a state was just changed, so it would be wasteful/almost dangerous).

I think this is the minimal set of function that can be done.

Reply:
sorry I meant, you could re-use `_onChanged()` in the `openEditor` function, when setting up the editors

Reply:
Indeed, doing that!